### PR TITLE
Fix CartRule : Discount on selected product excluding discounted products

### LINKED ISSUE #18065: Exclude discounted products: YES & Apply a discount to Selected product(s) Don't work correctly
#### Describe the bug

The coupon settings below do not work. The discount rate is incorrect because it also counts a discount on a promotional product. The product has a unique price (screen3).

#### Expected behavior

The coupon should give you a 790 HUF discount. 100% of the price of the non-promotional product. Or 50% if the coupon is set up this way.

-------------------------
>Edited by Khouloud

Steps to reproduce the issue:
1. Create a product `Product1` with a price = 36€ Tax included
2. Create a specific price on this product with 20€ Tax included
3. Create a second product Product2 with a price =120€ Tax included
4. Create a cart rule
4.1 With a code/ or not
4.2 Conditions => Product Selection ( add those product created Product1 & Product2 to the selection)
image
4.3 Actions: Apply a discount 100% to Selected products + Exclude discounted products enabled
image
5. Save
6. Go to the FO => add those products Product1 & Product2 to the cart
7. see error => Cart total should be calculated with cart rule applying only on the not-discounted product
image

This issue: https://github.com/PrestaShop/PrestaShop/issues/10300 is not fixed when we have the product selection in the cart rule.

------------------------------

#### Steps to Reproduce

**1. In the administration interface, add the coupon with the settings shown in the image.**
Screenshot_1
Screenshot_5
Screenshot_2

**2.  Add a specific price to a product**
Screenshot_4

**3. Put the products in to your cart:**
Screenshot_3

**Correctly, the coupon discount should only count on non-promotional products.**

#### Additional information

* PrestaShop version: 1.7.6.3
* PHP version: 7.2

The following reports and fixes not resolve this bug: #10454, #15876, #14407, #14442

### PR DESCRIPTION
**Pull request by Progi1984**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | 1.7.7.x
| Description?  | CartRule : Discount on selected product excluding discounted products
| Type?         | bug fix
| Category?     | CO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | Fixes #18065
| How to test?  | 1. Create a product `Product1` with a price = 36€ Tax included<br>2. Create a specific price on this product with 20€ Tax included<br>3. Create a second product Product2 with a price =120€ Tax included<br>4. Create a cart rule<br>4.1 With a code/ or not<br>4.2 Conditions => Product Selection ( add those product created Product1 & Product2 to the selection)<br>image<br>4.3 Actions: Apply a discount 100% to Selected products + Exclude discounted products enabled<br>image<br>5. Save<br>6. Go to the FO => add those products Product1 & Product2 to the cart<br>7. **BEFORE**see error => Cart total should be calculated with cart rule applying only on the not-discounted product<br>image<br>7. **AFTER** Cart total & Voucher price are correct<br>image

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/18144)

### CODE REVIEW

**Comment by matks** [comment]
> Hi!
> 
> Your pull request description seems to be incomplete or malformed:
> 
> * The `category` should be one of: `FO`, `BO`, `CO`, `IN`, `TE`, `WS`, `LO`, `ME`, `PM`
> 
> Would you mind completing the contribution table ? This would help us understand how interesting your contribution is.
> 
> Thank you!
> 
> (note: this is an automated message, but answering it will reach a real human )

@Progi1984 You must choose only one. If it's both BO and FO, then it's CO

**Review by matks — DISMISSED** [review]
Code style ✅ 
Code maintainability ✅ 

As for whether code is valid ... very hard to say in these complex classes. Then I will trust our behat tests

File: src/Adapter/Presenter/Cart/CartPresenter.php [hunk 1]
```diff
@@ -557,7 +558,22 @@ private function getTemplateVarVouchers(Cart $cart)
...
-                $productsTotalExcludingTax = $cart->getOrderTotal($this->includeTaxes(), Cart::ONLY_PRODUCTS);
+                $products = $cart->getProducts();
+                if ($this->cartVoucherHasPercentReductionOnSelectedProducts($cartVoucher)) {
+                    $selected_products = $cartVoucher['obj']->checkProductRestrictionsFromCart($cart, true);
+                    if (is_array($selected_products)) {
+                        foreach ($products as $key => $product) {
+                            if ((in_array($product['id_product'] . '-' . $product['id_product_attribute'], $selected_products)
+                                    || in_array($product['id_product'] . '-0', $selected_products))
+                                && (($cartVoucher['reduction_exclude_special'] && !$product['reduction_applies'])
+                                    || !$cartVoucher['reduction_exclude_special'])) {
```
Comment (matthieu-rolland) [inline src/Adapter/Presenter/Cart/CartPresenter.php:571]: Maybe it is worth adding a comment explaining this very complex condition ?

Comment (Progi1984) [reply src/Adapter/Presenter/Cart/CartPresenter.php:571]: Where is this complex :trollface: ?

Comment (Progi1984) [reply src/Adapter/Presenter/Cart/CartPresenter.php:571]: Comment added ;)

**Review by jolelievre — DISMISSED** [review]
I'm always amazed by the complexity of these cart rules.. And I have the feeling we mix everything in the Presenter And in the Calculator.. like the responsibility of each one is not clear..
But we're not gonna change this now anyway ^^

**Review by PierreRambaud — APPROVED** [review]

**Review by matthieu-rolland — APPROVED** [review]

**Comment by Robin-Fischer-PS** [comment]
Issue is fixed, and Sarah and I did some regression test : It's OK !

Thanks @Progi1984 !

**Comment by Progi1984** [comment]
Thanks @Robin-Fischer-PS

**Comment by jmm19** [comment]
I have Prestashop 1.7.6.2 
If I want to improve this bug in my shop without upgrade, is it enough change code in these two files above please? CartPresenter.php and CartRule.php?

**Pull request merged**